--- before/src/renderer.odin
+++ after/src/renderer.odin
@@ -51,7 +51,8 @@
 	defer wgpu.RenderPassEncoderRelease(pass)
 	wgpu.RenderPassEncoderSetPipeline(pass, renderer.pipeline)
 	wgpu.RenderPassEncoderSetVertexBuffer(pass, 0, renderer.vertex_buffer, 0, size_of(VERTICES))
-	wgpu.RenderPassEncoderDraw(pass, 3, 1, 0, 0)
+	wgpu.RenderPassEncoderSetIndexBuffer(pass, renderer.index_buffer, .Uint16, 0, size_of(INDICES))
+	wgpu.RenderPassEncoderDrawIndexed(pass, 6, 1, 0, 0, 0)
 	wgpu.RenderPassEncoderEnd(pass)
 	commands := wgpu.CommandEncoderFinish(
 		encoder,
@@ -76,15 +77,18 @@
 #assert(offset_of(Vertex, position) == 0)
 #assert(offset_of(Vertex, color) == 8)
 
-VERTICES :: [3]Vertex {
-	{{-0.6, -0.5}, {1, 0.1, 0.1, 1}},
-	{{0.6, -0.5}, {0.1, 1, 0.1, 1}},
-	{{0.0, 0.6}, {0.1, 0.1, 1, 1}},
+VERTICES :: [4]Vertex {
+	{{-0.25, 0.35}, {0.12, 0.78, 0.55, 1}},
+	{{0.25, 0.35}, {0.12, 0.78, 0.55, 1}},
+	{{0.25, -0.35}, {0.12, 0.78, 0.55, 1}},
+	{{-0.25, -0.35}, {0.12, 0.78, 0.55, 1}},
 }
+INDICES :: [6]u16{0, 1, 2, 0, 2, 3}
 
 Renderer :: struct {
 	pipeline:      wgpu.RenderPipeline,
 	vertex_buffer: wgpu.Buffer,
+	index_buffer:  wgpu.Buffer,
 }
 
 renderer_init :: proc(renderer: ^Renderer, gpu: ^GPU) -> bool {
@@ -96,6 +100,14 @@
 	if renderer.vertex_buffer == nil { return false }
 	vertices := VERTICES
 	wgpu.QueueWriteBuffer(gpu.queue, renderer.vertex_buffer, 0, &vertices, size_of(vertices))
+	renderer.index_buffer = wgpu.DeviceCreateBuffer(gpu.device, &wgpu.BufferDescriptor {
+			label = "quad indices",
+			size  = size_of(INDICES),
+			usage = {.Index, .CopyDst},
+		})
+	if renderer.index_buffer == nil { return false }
+	indices := INDICES
+	wgpu.QueueWriteBuffer(gpu.queue, renderer.index_buffer, 0, &indices, size_of(indices))
 	source := wgpu.ShaderSourceWGSL {
 		chain = {sType = .ShaderSourceWGSL},
 		code = #load("../shaders/character.wgsl", string),
@@ -150,6 +162,7 @@
 }
 
 renderer_destroy :: proc(renderer: ^Renderer) {
+	if renderer.index_buffer != nil { wgpu.BufferRelease(renderer.index_buffer) }
 	if renderer.vertex_buffer != nil { wgpu.BufferRelease(renderer.vertex_buffer) }
 	if renderer.pipeline != nil { wgpu.RenderPipelineRelease(renderer.pipeline) }
 	renderer^ = {}
